bionic/cache.py

```diff
@@ -1028,17 +1049,19 @@ def update_hash(hash_, obj):
 
 def relativize_url(artifact_url, metadata_url):
     """Returns the relative artifact url wrt to metadata url
-    when both urls start with file://. Otherwise, returns the original url."""
+    when both urls are file urls. Otherwise, returns the original url."""
     if not is_file_url(artifact_url) or not is_file_url(metadata_url):
         return artifact_url
-    return os.path.relpath(artifact_url, os.path.dirname(metadata_url))
+    artifact_path = path_from_url(artifact_url)
+    metadata_path = path_from_url(metadata_url)
+    return os.path.relpath(artifact_path, metadata_path.parent)
 
 
 def derelativize_url(artifact_url, metadata_url):
     """Returns the absolute artifact url when it is relative wrt metadata url.
     Otherwise returns the original url."""
-    if os.path.isabs(artifact_url):
+    if is_absolute_url(artifact_url):
         return artifact_url
-    abspath = os.path.normpath(os.path.join(os.path.dirname(metadata_url), artifact_url))
-    # normpath removes redundant `/`. We need to add them back from the url prefix
-    return abspath.replace("file:", FILE_URL_PREFIX)
+    metadata_path = path_from_url(metadata_url)
+    abspath = os.path.normpath(metadata_path.parent.joinpath(artifact_url))
```

Comment:
I know it doesn't change the behavior, but I feel like converting `artifact_url` to `artifact_path` first would make it more obvious that this code is safe.

Reply:
sure, if it looks safer